pkg/controller/persistentvolumelabel/pvlcontroller.go

```diff
@@ -0,0 +1,421 @@
+	if volume.Spec.GCEPersistentDisk.PDName == vol.ProvisionedVolumeName {
+		return nil, nil
+	}
+
+	provider, err := pvlc.getGCECloudProvider()
+	if err != nil {
+		return nil, err
+	}
+
+	// If the zone is already labeled, honor the hint
+	zone := volume.Labels[kubeletapis.LabelZoneFailureDomain]
+
+	labels, err := provider.GetAutoLabelsForPD(volume.Spec.GCEPersistentDisk.PDName, zone)
+	if err != nil {
+		return nil, err
+	}
+
+	return labels, nil
+}
+
+// getGCECloudProvider returns the GCE cloud provider, for use for querying volume labels
+func (pvlc *PersistentVolumeLabelController) getGCECloudProvider() (*gce.GCECloud, error) {
+	pvlc.mutex.Lock()
+	defer pvlc.mutex.Unlock()
+
+	if pvlc.gceCloudProvider == nil {
+		gceCloudProvider, ok := pvlc.cloud.(*gce.GCECloud)
+		if !ok {
+			// GetCloudProvider has gone very wrong
+			return nil, fmt.Errorf("error retrieving GCE cloud provider")
+		}
+		pvlc.gceCloudProvider = gceCloudProvider
+	}
+	return pvlc.gceCloudProvider, nil
+}
+
+func (pvlc *PersistentVolumeLabelController) updateVolume(volume *v1.PersistentVolume) error {
+	//func (pvlc *PersistentVolumeLabelController) updateVolume(labels map[string]string) error {
+	glog.V(4).Infof("updating PersistentVolume %s", volume.Name)
+	err := clientretry.RetryOnConflict(updatePVLabelsBackoff, func() error {
+		curVol, err := pvlc.kubeClient.Core().PersistentVolumes().Get(volume.Name, metav1.GetOptions{IncludeUninitialized: true})
+		if err != nil {
+			return err
+		}
+		oldData, err := json.Marshal(curVol)
+		if err != nil {
+			return fmt.Errorf("failed to marshal old persistentvolume %#v for persistentvolume %q: %v", curVol, volume.Name, err)
+		}
+
+		objCopy, err := scheme.Scheme.DeepCopy(curVol)
+		if err != nil {
+			return fmt.Errorf("failed to copy persistentvolume object %#v: %v", curVol, err)
+		}
+		newVolume, ok := (objCopy).(*v1.PersistentVolume)
+		if !ok {
+			return fmt.Errorf("failed to cast copy into persistentvolume object %#v: %v", newVolume, err)
+		}
+		newVolume.Labels = volume.Labels
+		pvlc.removeInitializer(newVolume)
+
+		newData, err := json.Marshal(newVolume)
+		if err != nil {
+			return fmt.Errorf("failed to marshal new persistentvolume %#v for persistentvolume %q: %v", newVolume, volume.Name, err)
+		}
+
+		patchBytes, err := strategicpatch.CreateTwoWayMergePatch(oldData, newData, v1.PersistentVolume{})
+		if err != nil {
+			return fmt.Errorf("failed to create patch for persistentvolume %q: %v", volume.Name, err)
+		}
+
+		_, err = pvlc.kubeClient.Core().PersistentVolumes().Patch(string(volume.Name), types.StrategicMergePatchType, patchBytes)
+		return err
+	})
+	if err == nil {
+		glog.V(4).Infof("updated PersistentVolume %s", volume.Name)
+	}
+
+	return err
+}
+
+func (pvlc *PersistentVolumeLabelController) addInitializationConfig() {
+	pvlInitializer := v1alpha1.Initializer{
+		Name: initializerName,
+		Rules: []v1alpha1.Rule{
+			{
+				APIGroups:   []string{"*"},
```

Comment:
At the time I thought I saw examples using "*", but looking at the e2e tests I see "" being used.  I'll change to ""